Fix Windows path resolution in "resolve_path" helper.

bennothommo opened the pull request:
Linked to #510 - this came up in the Windows unit tests. This PR should be merged in with that PR before merge.

Fixes https://github.com/octobercms/october/issues/5236.

Commit:
Fix Windows path resolution in "resolve_path" helper.

Fixes https://github.com/octobercms/october/issues/5236. Linked to https://github.com/octobercms/library/pull/510/.

Commit:
Update canonical path tests to work on both Windows and Linux

Commit:
Ignore normalisation test on Windows

Windows does in fact resolve Unix paths correctly through realpath().

Commit:
Handle Windows-based symlinks

Commit:
Include drive letter in assertions, if required

Commit:
Remove /tmp and /dev/null tests.

These directories do not always exist depending on the environment.

Commit:
Pass drive down to relative symlinks

Commit:
Correctly switch canon segments depending on OS after resolving symlink

Commit:
Correctly detect drive depending on path given, or current workdir

Commit:
Handle symlinks in tests depending on OS

LukeTowers commented on src/Support/helpers.php:
```diff
@@ -918,26 +918,31 @@ function title_case($value)
     {
         // Normalise directory separators
-        $path = str_replace('\\', '/', $path);
+        $path = str_replace(['\\', '/'], '/', $path);
```
What's the logic behind this change?

LukeTowers replied:
We might need to move this into a helper class that gets called from this function instead so that we can take advantage of methods to reduce / clarify some of the duplicated cleaning functionality

bennothommo replied:
@LukeTowers to make it easier to track and parse, I opted to change all directory separators while the path is being resolved to Unix `/` separators, regardless of OS. It's only when the final resolved path is generated and returned will the OS' directory separator be used.

bennothommo replied:
@LukeTowers I've converted the helper method to a Helper class (`October\Rain\Filesystem\PathResolver`) which might make it easier to track what is happening.

LukeTowers commented on src/Support/helpers.php:
```diff
@@ -918,26 +918,31 @@ function title_case($value)
+            ? $matches[1]
+            : null;
 
         // Check for a relative or absolute path, and prepend the working directory if relative, then split up the
         // path into segments
```
```suggestion
        // Check if this path is absolute or relative then split it into segments
```

LukeTowers commented on src/Support/helpers.php:
```diff
@@ -918,26 +918,31 @@ function title_case($value)
             $pathSegments = explode('/', $path);
         } else {
-            $path = getcwd() . '/' . $path;
+            $path = str_replace(['\\', '/'], '/', getcwd()) . '/' . $path;
```
```suggestion
            // Prepend the working directory to relative paths
            $path = str_replace(['\\', '/'], '/', getcwd()) . '/' . $path;
```

Commit:
Switch helper method to a PathResolver helper class.

Adds a new chained method "within" to determine if a given path is within a specified directory.

Commit:
Update tests

Commit:
Add comments

Commit:
Add newline at end of file

Commit:
Correctly handle symlink paths for Windows in new helper

bennothommo commented:
After discussing https://github.com/octobercms/october/issues/5236 with @Samuell1, we were able to determine that his fault also lies with the `starts_with` method call here: https://github.com/octobercms/october/blob/develop/modules/cms/classes/CmsObject.php#L235.

The path created in the lines before are using forward slashes as though they are a Unix path, but this won't be the case for Windows, so the `starts_with` method in conjunction with this `resolve_path` helper is failing on Windows. I have added a `within` chainable method to this new helper class which can normalise the given directory and determine if a path is within that directory.

LukeTowers commented on src/Filesystem/PathResolver.php:
```diff
@@ -0,0 +1,196 @@
+class PathResolver
+{
+    /** @var string The path to resolve or compare. */
+    protected $path;
```
I don't think this is worth being a class that needs to be instantiated, each method could very easily just be a static method, no need for state here.

bennothommo replied:
Yeah good point, I've simplified the class a bit.

Commit:
Simplify PathResolver helper class
